core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -2482,6 +2485,38 @@ private static void assertVisibleCount(InternalEngine engine, int numDocs, boole
         }
     }
 
+    public void testTranslogCleanUpPostCommitCrash() throws Exception {
+        try (Store store = createStore()) {
+            AtomicBoolean throwErrorOnCommit = new AtomicBoolean();
+            final Path translogPath = createTempDir();
+            try (InternalEngine engine = new InternalEngine(config(defaultSettings, store, translogPath, newMergePolicy(), null, null)) {
+                @Override
+                protected void commitIndexWriter(IndexWriter writer, Translog translog, String syncId) throws IOException {
```

Comment:
instead of making this method protected I think we should use `MockDirectoryWrapper#failOn(Failure)` and pass some failure to it that fails if we commit the indexwriter like this:
```Java
Failure fail = new Failure() {
      @Override
      public void eval(MockDirectoryWrapper dir) throws IOException {
        for (StackTraceElement e : Thread.currentThread().getStackTrace()) {
          if (doFail && "commit".equals(e.getMethodName())) {
            throw new FakeIOException();
          }
        }
      }
    };
```